### DISCUSSION TITLE
Performance measurements

### DISCUSSION THREAD

**Discussion by chaiiii12345** [open-edge-platform/anomalib#781] [Q&A] [answered]
Hi, I wanted to know if it's possible to get performance measurements like accuracy, precision, recall etc. other than the AUROC and F1Score? And since the latter two does require the use of recall, precision, etc. it should be possible to get them as output as well? I checked the codes in `/anomalib/utils/metrics` but I'm unsure how to do it 😅
Please help!!
I checked other discussions, and found #689, but it's unanswered...

**Comment by djdameln** [comment, marked as answer]
Anomalib uses the Torchmetrics package to evaluate model performance. Anomalib supports many metrics classes from the torchmetrics package out of the box, but also implements several custom metric implementations such as [AUROC](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/auroc.py#L17), [PRO](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/pro.py#L18) and [AUPRO](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/aupro.py#L23). 

The metrics that will be used to evaluate your models can be configured in the `metrics` section of the `config.yaml`. Image- and pixel-level metrics can be configured separately, respectively under `metrics.image` and `metrics.pixel`.

When adding a metric to the evaluation pipeline, Anomalib will first search for a metric with the specified name in the `anomalib.utils.metrics` module. If the metric is not found in Anomalib, it will look for the metric in the [available metrics of the Torchmetrics package](https://torchmetrics.readthedocs.io/en/stable/all-metrics.html). If no matching metrics are found in both Anomalib and Torchmetrics, a warning will be raised. The code responsible for this metric selection mechanism can be found [here](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/__init__.py#L26).

In your case, Precision, Recall and Accuracy are all available from the Torchmetrics package. So, to evaluate your Anomalib models using these metrics, simply list them under `metrics` in your `config.yaml`. Please note that the Accuracy metric will only work for the image-level evaluation, because the Torchmetrics [implementation](https://torchmetrics.readthedocs.io/en/stable/classification/accuracy.html) of this metric expects flattened inputs.

So to summarize, your desired use-case could be achieved by entering the following parameter values in the `metrics` section of your `config.yaml`.

```
metrics:
  image:
    - Precision
    - Recall
    - Accuracy
  pixel:
    - Precision
    - Recall
```

**Reply by chaiiii12345** [reply]
Oh that's awesome! Thank you!!

**Reply by nicozhou7** [reply]
can the adaptive threshold be calculated using the max Recall?

**Reply by djdameln** [reply]
> can the adaptive threshold be calculated using the max Recall?

No, the adaptive thresholding mechanism is obtained by maximizing F1 score over the validation set. The [AnomalyScoreThreshold](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/anomaly_score_threshold.py#L10) class is responsible for the computation. The adaptive thresholding mechanism can be disabled from the `config.yaml` by setting `metrics.threshold.method` to `manual`, but the metric class used for the computation cannot be changed. If you really want to use a different metric for the adaptive thresholding mechanism, you could of course replace the AnomalyScoreThreshold class with your own custom implementation.

In any case, using Recall to compute the adaptive threshold is a bad idea. Recall, or true positive rate, quantifies how many true positives (in this case: anomalous) are retrieved with respect to the total number of items in the dataset. So false positives are not penalized by the recall metric. This means that you can always get an equal or better recall score by lowering your threshold. When used for adaptive thresholding, this would effectively mean that the threshold is lowered until all anomalous objects have been retrieved and the recall is 1.0. Since false positives are not penalized, it is very likely that the obtained threshold value will lead to many false positives, so its value will not be useful.

**Reply by nicozhou7** [reply]
Thanks @djdameln for the explanation! Really appreciate

**Reply by nicozhou7** [reply]
Btw, I am still confused of where does the anomalib do the computation to get the maximal pixel distance from good data for threshold if we do not has the defective image for validation

**Reply by djdameln** [reply]
I just added a brief answer to this question in the other thread: https://github.com/openvinotoolkit/anomalib/discussions/192#discussioncomment-4424354

**Reply by electro020** [reply]
> Anomalib uses the Torchmetrics package to evaluate model performance. Anomalib supports many metrics classes from the torchmetrics package out of the box, but also implements several custom metric implementations such as [AUROC](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/auroc.py#L17), [PRO](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/pro.py#L18) and [AUPRO](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/aupro.py#L23).
> 
> The metrics that will be used to evaluate your models can be configured in the `metrics` section of the `config.yaml`. Image- and pixel-level metrics can be configured separately, respectively under `metrics.image` and `metrics.pixel`.
> 
> When adding a metric to the evaluation pipeline, Anomalib will first search for a metric with the specified name in the `anomalib.utils.metrics` module. If the metric is not found in Anomalib, it will look for the metric in the [available metrics of the Torchmetrics package](https://torchmetrics.readthedocs.io/en/stable/all-metrics.html). If no matching metrics are found in both Anomalib and Torchmetrics, a warning will be raised. The code responsible for this metric selection mechanism can be found [here](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/__init__.py#L26).
> 
> In your case, Precision, Recall and Accuracy are all available from the Torchmetrics package. So, to evaluate your Anomalib models using these metrics, simply list them under `metrics` in your `config.yaml`. Please note that the Accuracy metric will only work for the image-level evaluation, because the Torchmetrics [implementation](https://torchmetrics.readthedocs.io/en/stable/classification/accuracy.html) of this metric expects flattened inputs.
> 
> So to summarize, your desired use-case could be achieved by entering the following parameter values in the `metrics` section of your `config.yaml`.
> 
> ```
> metrics:
>   image:
>     - Precision
>     - Recall
>     - Accuracy
>   pixel:
>     - Precision
>     - Recall
> ```

Hello,
Is there a similar way to obtain the loss curve during the training ?
I would like to have a graph

**Reply by ashwinvaidya17** [reply]
@electro020 we just merged a [PR](https://github.com/openvinotoolkit/anomalib/pull/804) to log the loss. The [lightning documentation](https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#logging-from-a-lightningmodule) will be useful to you.

**Reply by electro020** [reply]
> @electro020 we just merged a [PR](https://github.com/openvinotoolkit/anomalib/pull/804) to log the loss. The [lightning documentation](https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#logging-from-a-lightningmodule) will be useful to you.

Hello, 
When i looked at the log ,  i couldnt see the loss:
![image](https://user-images.githubusercontent.com/99293983/209249136-e2985615-03ac-4299-81f9-719b2d018d69.png)
Is it supposed to appear there ?

thank you
